File: 8vim/src/main/java/inc/flide/vim8/ui/fragments/LayoutFileSelector.java
package inc.flide.vim8.ui.fragments;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.preference.PreferenceFragmentCompat;
import arrow.core.EitherKt;
import inc.flide.vim8.R;
import inc.flide.vim8.keyboardactionlisteners.MainKeypadActionListener;
import inc.flide.vim8.models.AppPrefs;
import inc.flide.vim8.models.CustomLayout;
import inc.flide.vim8.models.LayoutKt;
import inc.flide.vim8.models.error.ExceptionWrapperError;
import inc.flide.vim8.utils.DialogsHelper;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import kotlin.Pair;

public abstract class LayoutFileSelector extends PreferenceFragmentCompat {
    private static final String[] LAYOUT_FILTER = { "application/octet-stream" };

    protected AppPrefs prefs;

    private final ActivityResultLauncher<String[]> openContent = registerForActivityResult(
            new ActivityResultContracts.OpenDocument(), this::callback);

    private void callback(Uri selectedCustomLayoutFile) {
        AppPrefs.Layout layoutPrefs = prefs.getLayout();
        Context context = getContext();
        if (selectedCustomLayoutFile == null || context == null) {
            return;
        }
        final int takeFlags = Intent.FLAG_GRANT_READ_URI_PERMISSION;
        context.getContentResolver()
                .takePersistableUriPermission(selectedCustomLayoutFile, takeFlags);
        CustomLayout layout = new CustomLayout(selectedCustomLayoutFile);
        Set<String> currentHistory = layoutPrefs.getCustom().getHistory().get();
        boolean isInHistory = currentHistory.contains(selectedCustomLayoutFile.toString());
        if (isInHistory) {
            updateKeyboard(layout, context);
        } else {
            Pair<Integer, String> errorToShow = EitherKt.getOrElse(
                    LayoutKt
                            .loadKeyboardData(layout, context)
                            .map(keyboardData -> {
                                if (keyboardData.getTotalLayers() == 0) {
                                    return new Pair<>(R.string.yaml_error_title,
                                            "The layout requires at least one layer");
                                }
                                return null;
                            }),
                    error -> {
                        int title = R.string.yaml_error_title;
                        if (error instanceof ExceptionWrapperError) {
                            title = R.string.generic_error_text;
                        }
                        return new Pair<>(title, error.getMessage());
                    });
            if (errorToShow != null) {
                DialogsHelper.showAlert(context, errorToShow.getFirst(), errorToShow.getSecond());
                return;
            }
            updateKeyboard(layout, context);
            List<String> history = new ArrayList<>(currentHistory);
            history.add(0, selectedCustomLayoutFile.toString());
            layoutPrefs.getCustom().getHistory().set(new LinkedHashSet<>(history), true);
        }
    }

    private void updateKeyboard(CustomLayout layout, Context context) {
        prefs.getLayout().getCurrent().set(layout, true);
        MainKeypadActionListener.rebuildKeyboardData(LayoutKt.loadKeyboardData(layout, context).getOrNull());
    }

    protected void openFileSelector() {
        openContent.launch(LAYOUT_FILTER);
    }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/app/settings/LayoutFileSelector.kt
package inc.flide.vim8.app.settings

import android.content.Intent
import android.net.Uri
import androidx.activity.result.contract.ActivityResultContracts
import androidx.preference.PreferenceFragmentCompat
import arrow.core.None
import arrow.core.some
import inc.flide.vim8.R
import inc.flide.vim8.appPreferenceModel
import inc.flide.vim8.ime.LayoutLoader
import inc.flide.vim8.ime.layout.AvailableLayouts
import inc.flide.vim8.ime.layout.CustomLayout
import inc.flide.vim8.ime.layout.loadKeyboardData
import inc.flide.vim8.ime.layout.models.KeyboardData
import inc.flide.vim8.ime.layout.models.error.ExceptionWrapperError
import inc.flide.vim8.ime.layout.models.error.LayoutError
import inc.flide.vim8.utils.DialogsHelper.showAlert

abstract class LayoutFileSelector : PreferenceFragmentCompat() {
    protected val prefs by appPreferenceModel()
    protected var availableLayouts: AvailableLayouts? = null
    protected var layoutLoader: LayoutLoader? = null
    private val openContent = registerForActivityResult<Array<String>, Uri>(
        ActivityResultContracts.OpenDocument()
    ) { selectedCustomLayoutFile: Uri? -> callback(selectedCustomLayoutFile) }

    private fun callback(selectedCustomLayoutFile: Uri?) {
        val layoutPrefs = prefs.layout
        val context = context
        if (selectedCustomLayoutFile == null || context == null) {
            return
        }
        context.contentResolver
            .takePersistableUriPermission(
                selectedCustomLayoutFile,
                Intent.FLAG_GRANT_READ_URI_PERMISSION
            )
        val layout = CustomLayout(selectedCustomLayoutFile)
        val currentHistory = layoutPrefs.custom.history.get()
        val isInHistory = currentHistory.contains(selectedCustomLayoutFile.toString())
        if (isInHistory) {
            availableLayouts!!.updateKeyboardData(layout)
        } else {
            layout.loadKeyboardData(layoutLoader!!, context)
                .fold({ error: LayoutError ->
                    val title = if (error is ExceptionWrapperError) {
                        R.string.generic_error_text
                    } else {
                        R.string.yaml_error_title
                    }
                    (title to error.message).some()
                }, { keyboardData: KeyboardData ->
                    if (keyboardData.totalLayers == 0) {
                        (R.string.yaml_error_title to "The layout requires at least one layer")
                            .some()
                    } else {
                        None
                    }
                })
                .onSome { (titleId, message) -> showAlert(context, titleId, message) }
                .onNone {
                    prefs.layout.current.set(layout)
                    val history =
                        listOf(selectedCustomLayoutFile.toString()) + currentHistory.toList()

                    layoutPrefs.custom.history.set(LinkedHashSet(history))
                }
        }
    }

    protected fun openFileSelector() {
        openContent.launch(LAYOUT_FILTER)
    }

    companion object {
        private val LAYOUT_FILTER = arrayOf("application/octet-stream")
    }
}
